File: tif_customization/tif_customization/page/school_wise_demand_report/school_wise_demand_report.js
frappe.pages["school-wise-demand-report"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("School Wise Demand Report"),
		single_column: true,
	});
	new frappe.tif_customization.SchoolWiseDemandReport(page).make();
};

frappe.tif_customization = frappe.tif_customization || {};

frappe.tif_customization.SchoolWiseDemandReport = class SchoolWiseDemandReport {
	constructor(page) {
		this.page = page;
		this.rows = [];
		this.summary = {};
		this._load_timer = null;
	}

	make() {
		this.make_layout();
		this.make_filters();
		this.page.set_primary_action(__("Refresh"), () => this.load_data(), "refresh");
		this.page.add_action_item(__("Export CSV"), () => this.export_csv());
		this.page.add_action_item(__("Print"), () => this.print_report());
		this.load_data();
	}

	make_layout() {
		$(this.page.body).html(`
			<div style="padding:16px;">
				<style>
					.swdr-kpis{display:flex;flex-wrap:wrap;gap:12px;margin-bottom:16px;align-items:stretch}
					.swdr-kpi-col{flex:1 1 150px;min-width:140px;max-width:100%;display:flex}
					.swdr-kpi{
						position:relative;overflow:hidden;
						padding:14px 16px;border-radius:10px;
						border:1px solid var(--border-color,#e5e7eb);
						width:100%;
						min-height:108px;
						height:100%;
						display:flex;flex-direction:column;justify-content:space-between;
						box-sizing:border-box;
					}
					.swdr-kpi::before{
						content:"";position:absolute;inset:0 0 auto 0;height:4px;
						background:var(--swdr-accent, transparent);
					}
					.swdr-kpi .label{font-size:12px;margin:0 0 6px;line-height:1.3;color:var(--swdr-muted, var(--text-muted,#6b7280))}
					.swdr-kpi .value{font-size:22px;font-weight:700;line-height:1.2;color:var(--swdr-value, var(--text-color,#111827))}
					.swdr-kpi .hint{font-size:11px;margin-top:auto;padding-top:6px;line-height:1.3;color:var(--swdr-muted, var(--text-muted,#6b7280))}
					.swdr-kpi--schools{--swdr-accent:#1e40af;--swdr-value:#fff;--swdr-muted:#dbeafe;background:linear-gradient(135deg,#1d4ed8 0%,#1e3a8a 100%);border-color:#1e40af;box-shadow:0 6px 16px rgba(30,64,175,.18)}
					.swdr-kpi--orders{--swdr-accent:#0f766e;--swdr-value:#fff;--swdr-muted:#ccfbf1;background:linear-gradient(135deg,#0d9488 0%,#115e59 100%);border-color:#0f766e;box-shadow:0 6px 16px rgba(15,118,110,.18)}
					.swdr-kpi--pending{--swdr-accent:#ea580c;--swdr-value:#c2410c;--swdr-muted:#7c2d12;background:linear-gradient(135deg,#fff7ed 0%,#ffedd5 100%);border-color:#fed7aa}
					.swdr-kpi--mqh{--swdr-accent:#7c3aed;--swdr-value:#6d28d9;--swdr-muted:#4c1d95;background:linear-gradient(135deg,#f5f3ff 0%,#ede9fe 100%);border-color:#ddd6fe}
					.swdr-kpi--qaida{--swdr-accent:#059669;--swdr-value:#047857;--swdr-muted:#064e3b;background:linear-gradient(135deg,#ecfdf5 0%,#d1fae5 100%);border-color:#a7f3d0}
					.swdr-kpi--guide{--swdr-accent:#dc2626;--swdr-value:#b91c1c;--swdr-muted:#7f1d1d;background:linear-gradient(135deg,#fef2f2 0%,#fee2e2 100%);border-color:#fecaca}
					.swdr-note{font-size:12px;color:var(--text-muted,#6b7280);margin:0 0 12px}
					.swdr-table-wrap{overflow:auto;border:1px solid var(--border-color,#e5e7eb);border-radius:8px;background:#fff}
					.swdr-table{width:100%;border-collapse:collapse;font-size:13px}
					.swdr-table th{background:var(--control-bg,#f4f5f6);text-align:left;padding:10px 12px;border-bottom:1px solid var(--border-color,#e5e7eb);white-space:nowrap}
					.swdr-table td{padding:9px 12px;border-bottom:1px solid var(--border-color,#eee);vertical-align:top}
					.swdr-table tr:hover td{background:#fafafa}
					.swdr-num{text-align:right;font-variant-numeric:tabular-nums}
					.swdr-muted{color:var(--text-muted,#6b7280)}
					.swdr-pending{color:#b45309;font-weight:600}
				</style>
				<p class="swdr-note">
					Demand = open Sales Orders not fully delivered (book items only).
					Pending qty = Ordered − Delivered. Use School Name to filter by customer / school.
				</p>
				<div id="swdr-filters" class="row" style="margin-bottom:12px;"></div>
				<div id="swdr-kpis" class="swdr-kpis"></div>
				<div id="swdr-table" class="swdr-table-wrap"></div>
			</div>
		`);
	}

	make_filters() {
		this.from_date = this.make_filter({
			label: __("Order From Date"),
			fieldtype: "Date",
			fieldname: "from_date",
		});
		this.to_date = this.make_filter({
			label: __("Order To Date"),
			fieldtype: "Date",
			fieldname: "to_date",
		});
		this.customer = this.make_filter({
			label: __("School Name"),
			fieldtype: "Link",
			fieldname: "customer",
			options: "Customer",
		});
		this.school_name = this.make_filter({
			label: __("School Name Contains"),
			fieldtype: "Data",
			fieldname: "school_name",
		});
		this.school = this.make_filter({
			label: __("School (Master)"),
			fieldtype: "Link",
			fieldname: "school",
			options: "School",
		});
		this.city = this.make_filter({
			label: __("City"),
			fieldtype: "Link",
			fieldname: "city",
			options: "City",
		});
		this.province = this.make_filter({
			label: __("Province"),
			fieldtype: "Data",
			fieldname: "province",
		});
		this.area = this.make_filter({
			label: __("Area"),
			fieldtype: "Data",
			fieldname: "area",
		});
		this.school_status = this.make_filter({
			label: __("School Status"),
			fieldtype: "Select",
			fieldname: "school_status",
			options:
				"\nActive\nInactive\nIn Process\nClosed\nNot Interested\nDirect Requirement Received",
		});
		this.book_type = this.make_filter({
			label: __("Book Type"),
			fieldtype: "Select",
			fieldname: "book_type",
			options: "\nMQH\nQaida\nGuide",
		});
	}

	make_filter(df) {
		const wrap = $(`<div class="col-md-3 col-sm-6" style="margin-bottom:8px;"></div>`).appendTo(
			$(this.page.body).find("#swdr-filters")
		);
		const control = frappe.ui.form.make_control({
			df: Object.assign({}, df, {
				change: () => this.schedule_load(),
			}),
			parent: wrap,
			render_input: true,
		});
		control.refresh();
		if (df.default != null) {
			control.set_value(df.default);
		}
		control.$input?.on("awesomplete-selectcomplete", () => this.schedule_load());
		control.$input?.on("keydown", (e) => {
			if (e.which === 13) this.schedule_load();
		});
		return control;
	}

	schedule_load() {
		clearTimeout(this._load_timer);
		this._load_timer = setTimeout(() => this.load_data(), 250);
	}

	get_filters() {
		return {
			from_date: this.from_date.get_value(),
			to_date: this.to_date.get_value(),
			customer: this.customer.get_value(),
			school_name: this.school_name.get_value(),
			school: this.school.get_value(),
			city: this.city.get_value(),
			province: this.province.get_value(),
			area: this.area.get_value(),
			school_status: this.school_status.get_value(),
			book_type: this.book_type.get_value(),
		};
	}

	async load_data() {
		$(this.page.body).find("#swdr-table").html(`<div class="swdr-muted" style="padding:16px;">Loading…</div>`);
		const r = await frappe.call({
			method:
				"tif_customization.tif_customization.page.school_wise_demand_report.school_wise_demand_report.get_report_data",
			args: { filters: this.get_filters() },
		});
		this.rows = r.message?.rows || [];
		this.summary = r.message?.summary || {};
		this.render_kpis();
		this.render_table();
	}

	render_kpis() {
		const s = this.summary;
		const cards = [
			{ label: __("Schools"), value: s.schools || 0, cls: "swdr-kpi--schools", hint: __("With open demand") },
			{ label: __("Open Sales Orders"), value: s.sales_orders || 0, cls: "swdr-kpi--orders", hint: __("Not fully delivered") },
			{ label: __("Total Pending"), value: s.total_pending || 0, cls: "swdr-kpi--pending", hint: __("Books still due") },
			{ label: __("MQH Pending"), value: s.mqh_pending || 0, cls: "swdr-kpi--mqh", hint: __("MQH demand") },
			{ label: __("Qaida Pending"), value: s.qaida_pending || 0, cls: "swdr-kpi--qaida", hint: __("Qaida demand") },
			{ label: __("Guide Pending"), value: s.guide_pending || 0, cls: "swdr-kpi--guide", hint: __("Guide demand") },
		];
		$(this.page.body)
			.find("#swdr-kpis")
			.html(
				cards
					.map(
						(c) => `
					<div class="swdr-kpi-col">
						<div class="swdr-kpi ${c.cls}">
							<div class="label">${frappe.utils.escape_html(c.label)}</div>
							<div class="value">${frappe.format(c.value, { fieldtype: "Float", precision: 0 })}</div>
							<div class="hint">${frappe.utils.escape_html(c.hint || "")}</div>
						</div>
					</div>`
					)
					.join("")
			);
	}

	columns() {
		return [
			{ key: "school_name", label: __("School") },
			{ key: "school_status", label: __("Status") },
			{ key: "city", label: __("City") },
			{ key: "province", label: __("Province") },
			{ key: "area", label: __("Area") },
			{ key: "sales_orders", label: __("Open SOs"), num: true },
			{ key: "total_ordered", label: __("Ordered"), num: true },
			{ key: "total_delivered", label: __("Delivered"), num: true },
			{ key: "total_pending", label: __("Pending"), num: true, pending: true },
			{ key: "mqh_pending", label: __("MQH Pend"), num: true, pending: true },
			{ key: "qaida_pending", label: __("Qaida Pend"), num: true, pending: true },
			{ key: "guide_pending", label: __("Guide Pend"), num: true, pending: true },
			{ key: "other_pending", label: __("Other Pend"), num: true, pending: true },
			{ key: "last_order_date", label: __("Last Order") },
			{ key: "earliest_delivery_date", label: __("Earliest Delivery") },
		];
	}

	render_table() {
		const cols = this.columns();
		if (!this.rows.length) {
			$(this.page.body)
				.find("#swdr-table")
				.html(`<div class="swdr-muted" style="padding:16px;">No open school demand for selected filters.</div>`);
			return;
		}
		const head = cols
			.map((c) => `<th class="${c.num ? "swdr-num" : ""}">${frappe.utils.escape_html(c.label)}</th>`)
			.join("");
		const body = this.rows
			.map((row) => {
				const cells = cols
					.map((c) => {
						let val = row[c.key];
						let cls = c.num ? "swdr-num" : "";
						if (c.num) {
							const num = Number(val || 0);
							val = frappe.format(num, { fieldtype: "Float", precision: 0 });
							if (c.pending && num > 0) cls += " swdr-pending";
						} else if ((c.key === "last_order_date" || c.key === "earliest_delivery_date") && val) {
							val = frappe.datetime.str_to_user(val);
						} else {
							val = frappe.utils.escape_html(val || "");
						}
						return `<td class="${cls}">${val}</td>`;
					})
					.join("");
				return `<tr>${cells}</tr>`;
			})
			.join("");
		$(this.page.body).find("#swdr-table").html(`
			<table class="swdr-table">
				<thead><tr>${head}</tr></thead>
				<tbody>${body}</tbody>
			</table>
		`);
	}

	export_csv() {
		if (!this.rows.length) {
			frappe.msgprint(__("No data to export"));
			return;
		}
		const cols = this.columns();
		const lines = [cols.map((c) => c.label).join(",")];
		this.rows.forEach((row) => {
			lines.push(
				cols
					.map((c) => {
						const v = row[c.key] == null ? "" : String(row[c.key]);
						return `"${v.replace(/"/g, '""')}"`;
					})
					.join(",")
			);
		});
		const blob = new Blob([lines.join("\n")], { type: "text/csv;charset=utf-8;" });
		const a = document.createElement("a");
		a.href = URL.createObjectURL(blob);
		a.download = `school_wise_demand_report_${frappe.datetime.get_today()}.csv`;
		a.click();
	}

	print_report() {
		const html = $(this.page.body).find("#swdr-table").html();
		const w = window.open("", "_blank");
		w.document.write(`
			<html><head><title>School Wise Demand Report</title>
			<style>
				body{font-family:sans-serif;padding:16px}
				table{width:100%;border-collapse:collapse;font-size:12px}
				th,td{border:1px solid #ddd;padding:6px 8px}
				th{background:#f4f5f6;text-align:left}
				.swdr-num{text-align:right}
			</style></head><body>
			<h2>School Wise Demand Report</h2>
			${html}
			</body></html>
		`);
		w.document.close();
		w.focus();
		w.print();
	}
};

File: tif_customization/tif_customization/page/school_wise_demand_report/school_wise_demand_report.py
# Copyright (c) 2026, TIF Customization and contributors
# License: MIT

import json

import frappe
from frappe.utils import flt, getdate

from tif_customization.tif_customization.utils.supply_chain_books import (
	sql_book_item_filter,
	with_book_item_params,
)
from tif_customization.tif_customization.page.school_wise_book_dispatched.school_wise_book_dispatched import (
	_apply_school_name_filter,
)


@frappe.whitelist()
def get_report_data(filters=None):
	filters = _parse_filters(filters)
	rows = _get_school_demand_rows(filters)
	return {"rows": rows, "summary": _build_summary(rows)}


def _parse_filters(filters):
	if isinstance(filters, str):
		filters = json.loads(filters or "{}")
	return filters or {}


def _get_school_demand_rows(filters):
	where = [
		"so.docstatus = 1",
		"so.status NOT IN ('Closed', 'Cancelled', 'Completed')",
		"IFNULL(so.per_delivered, 0) < 100",
		"(IFNULL(soi.qty, 0) - IFNULL(soi.delivered_qty, 0)) > 0",
		sql_book_item_filter("soi.item_code").lstrip("AND ").strip(),
	]
	params = with_book_item_params()

	if filters.get("from_date"):
		where.append("so.transaction_date >= %(from_date)s")
		params["from_date"] = getdate(filters["from_date"])
	if filters.get("to_date"):
		where.append("so.transaction_date <= %(to_date)s")
		params["to_date"] = getdate(filters["to_date"])

	if filters.get("school") or filters.get("school_name") or filters.get("customer"):
		_apply_school_name_filter(
			where,
			params,
			filters,
			customer_field="so.customer",
			customer_name_field="so.customer_name",
		)

	if filters.get("city"):
		where.append("(COALESCE(s.city, addr.city, '') = %(city)s)")
		params["city"] = filters["city"]

	if filters.get("province"):
		where.append(
			"(TRIM(COALESCE(s.province, addr.state, '')) = %(province)s)"
		)
		params["province"] = str(filters["province"]).strip()

	if filters.get("sales_order"):
		where.append("so.name = %(sales_order)s")
		params["sales_order"] = filters["sales_order"]

	if filters.get("area"):
		where.append("(COALESCE(s.area, addr.custom_area, '') = %(area)s)")
		params["area"] = filters["area"]

	if filters.get("school_status"):
		where.append("s.status = %(school_status)s")
		params["school_status"] = filters["school_status"]

	book_type = (filters.get("book_type") or "").strip()
	if book_type == "MQH":
		where.append(
			"(UPPER(COALESCE(soi.item_name,'')) LIKE '%%MQH%%' OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%MQH%%')"
		)
	elif book_type == "Qaida":
		where.append(
			"(UPPER(COALESCE(soi.item_name,'')) LIKE '%%QAIDA%%' OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%QAIDA%%')"
		)
	elif book_type == "Guide":
		where.append(
			"(UPPER(COALESCE(soi.item_name,'')) LIKE '%%GUIDE%%' OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%GUIDE%%')"
		)
	elif book_type == "KPK Edition":
		where.append(_kpk_edition_sql())

	where_clause = " AND ".join(where)
	rows = frappe.db.sql(
		f"""
		SELECT
			COALESCE(s.name, '') AS school,
			COALESCE(NULLIF(s.school_name, ''), so.customer_name, so.customer) AS school_name,
			so.customer AS customer,
			COALESCE(s.status, '') AS school_status,
			COALESCE(s.school_type, '') AS school_type,
			COALESCE(s.city, addr.city, '') AS city,
			TRIM(COALESCE(s.province, addr.state, '')) AS province,
			COALESCE(s.area, addr.custom_area, '') AS area,
			COUNT(DISTINCT so.name) AS sales_orders,
			GROUP_CONCAT(DISTINCT so.name ORDER BY so.transaction_date DESC, so.name SEPARATOR ', ') AS sales_order_nos,
			SUM(IFNULL(soi.qty, 0) - IFNULL(soi.delivered_qty, 0)) AS total_pending,
			SUM(IFNULL(soi.qty, 0)) AS total_ordered,
			SUM(IFNULL(soi.delivered_qty, 0)) AS total_delivered,
			SUM(
				CASE
					WHEN UPPER(COALESCE(soi.item_name,'')) LIKE '%%MQH%%'
						OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%MQH%%'
					THEN IFNULL(soi.qty, 0) - IFNULL(soi.delivered_qty, 0) ELSE 0
				END
			) AS mqh_pending,
			SUM(
				CASE
					WHEN UPPER(COALESCE(soi.item_name,'')) LIKE '%%QAIDA%%'
						OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%QAIDA%%'
					THEN IFNULL(soi.qty, 0) - IFNULL(soi.delivered_qty, 0) ELSE 0
				END
			) AS qaida_pending,
			SUM(
				CASE
					WHEN UPPER(COALESCE(soi.item_name,'')) LIKE '%%GUIDE%%'
						OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%GUIDE%%'
					THEN IFNULL(soi.qty, 0) - IFNULL(soi.delivered_qty, 0) ELSE 0
				END
			) AS guide_pending,
			SUM(
				CASE
					WHEN {_kpk_edition_case()}
					THEN IFNULL(soi.qty, 0) - IFNULL(soi.delivered_qty, 0) ELSE 0
				END
			) AS kpk_pending,
			MAX(so.transaction_date) AS last_order_date,
			MIN(so.delivery_date) AS earliest_delivery_date
		FROM `tabSales Order` so
		INNER JOIN `tabSales Order Item` soi ON soi.parent = so.name
		LEFT JOIN `tabAddress` addr ON addr.name = so.customer_address
		LEFT JOIN `tabSchool` s ON (
			s.school_name = so.customer_name
			OR s.name = so.customer
		)
		WHERE {where_clause}
		GROUP BY
			COALESCE(s.name, ''),
			COALESCE(NULLIF(s.school_name, ''), so.customer_name, so.customer),
			so.customer,
			COALESCE(s.status, ''),
			COALESCE(s.school_type, ''),
			COALESCE(s.city, addr.city, ''),
			TRIM(COALESCE(s.province, addr.state, '')),
			COALESCE(s.area, addr.custom_area, '')
		ORDER BY total_pending DESC, school_name ASC
		""",
		params,
		as_dict=True,
	)

	for row in rows:
		for key in (
			"total_pending",
			"total_ordered",
			"total_delivered",
			"mqh_pending",
			"qaida_pending",
			"guide_pending",
			"kpk_pending",
		):
			row[key] = flt(row.get(key))
		row["other_pending"] = max(
			0.0,
			flt(
				row["total_pending"]
				- row["mqh_pending"]
				- row["qaida_pending"]
				- row["guide_pending"]
				- row["kpk_pending"]
			),
		)
		row["sales_orders"] = int(row.get("sales_orders") or 0)
		row["sales_order_nos"] = (row.get("sales_order_nos") or "").strip()
		row["province"] = (row.get("province") or "").strip()
		if row.get("last_order_date"):
			row["last_order_date"] = str(getdate(row["last_order_date"]))
		if row.get("earliest_delivery_date"):
			row["earliest_delivery_date"] = str(getdate(row["earliest_delivery_date"]))

	return rows


def _kpk_edition_sql():
	"""Match KPK Edition / KPK Textbook Board book lines."""
	return (
		"(UPPER(COALESCE(soi.item_name,'')) LIKE '%%KPK%%' "
		"OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%KPK%%')"
	)


def _kpk_edition_case():
	return (
		"UPPER(COALESCE(soi.item_name,'')) LIKE '%%KPK%%' "
		"OR UPPER(COALESCE(soi.item_code,'')) LIKE '%%KPK%%'"
	)


PROVINCE_OPTIONS = (
	"Sindh",
	"Punjab",
	"KPK",
	"Balochistan",
	"ICT",
	"AJK",
	"Gilgit Baltistan",
)


@frappe.whitelist()
def get_filter_options():
	"""Province list for filter (always includes KPK + values seen on open demand)."""
	seen = set(
		frappe.db.sql(
			"""
			SELECT DISTINCT TRIM(COALESCE(s.province, addr.state, '')) AS province
			FROM `tabSales Order` so
			LEFT JOIN `tabAddress` addr ON addr.name = so.customer_address
			LEFT JOIN `tabSchool` s ON (
				s.school_name = so.customer_name
				OR s.name = so.customer
			)
			WHERE so.docstatus = 1
			  AND so.status NOT IN ('Closed', 'Cancelled', 'Completed')
			  AND IFNULL(so.per_delivered, 0) < 100
			  AND TRIM(COALESCE(s.province, addr.state, '')) != ''
			"""
		)
	)
	provinces = []
	for p in PROVINCE_OPTIONS:
		provinces.append(p)
	for p in sorted(v[0] for v in seen if v and v[0]):
		if p not in provinces:
			provinces.append(p)
	return {"provinces": provinces}


def _build_summary(rows):
	return {
		"schools": len(rows),
		"sales_orders": sum(int(r.get("sales_orders") or 0) for r in rows),
		"total_pending": sum(flt(r.get("total_pending")) for r in rows),
		"mqh_pending": sum(flt(r.get("mqh_pending")) for r in rows),
		"qaida_pending": sum(flt(r.get("qaida_pending")) for r in rows),
		"guide_pending": sum(flt(r.get("guide_pending")) for r in rows),
		"kpk_pending": sum(flt(r.get("kpk_pending")) for r in rows),
		"total_ordered": sum(flt(r.get("total_ordered")) for r in rows),
		"total_delivered": sum(flt(r.get("total_delivered")) for r in rows),
	}
